Stripe/STPSourcePoller.m

```diff
@@ -0,0 +1,209 @@
+//
+//  STPSourcePoller.m
+//  Stripe
+//
+//  Created by Ben Guo on 1/26/17.
+//  Copyright © 2017 Stripe, Inc. All rights reserved.
+//
+
+#import "STPAPIClient+Private.h"
+#import "STPAPIRequest.h"
+#import "STPSource.h"
+#import "STPSourcePoller.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+static NSTimeInterval const DefaultPollInterval = 1.5;
+static NSTimeInterval const MaxPollInterval = 24;
+// Stop polling after 5 minutes
+static NSTimeInterval const Timeout = 60*5;
+// Stop polling after 5 consecutive non-200 responses
+static NSTimeInterval const MaxRetries = 5;
+
+@interface STPSourcePoller ()
+
+@property (nonatomic, weak) STPAPIClient *apiClient;
+@property (nonatomic) NSString *sourceID;
+@property (nonatomic) NSString *clientSecret;
+@property (nonatomic, copy) STPSourceCompletionBlock completion;
+@property (nonatomic, nullable) STPSource *latestSource;
+@property (nonatomic) NSTimeInterval pollInterval;
+@property (nonatomic, nullable) NSURLSessionDataTask *dataTask;
+@property (nonatomic, nullable) NSTimer *timer;
+@property (nonatomic) NSDate *startTime;
+@property (nonatomic) NSInteger retryCount;
+@property (nonatomic) NSInteger requestCount;
+@property (nonatomic) BOOL pollingPaused;
+@property (nonatomic) BOOL pollingStopped;
+
+@end
+
+@implementation STPSourcePoller
+
+- (instancetype)initWithAPIClient:(STPAPIClient *)apiClient
+                     clientSecret:(NSString *)clientSecret
+                         sourceID:(NSString *)sourceID
+                       completion:(STPSourceCompletionBlock)completion {
+    self = [super init];
+    if (self) {
+        _apiClient = apiClient;
+        _sourceID = sourceID;
+        _clientSecret = clientSecret;
+        _completion = completion;
+        _pollInterval = DefaultPollInterval;
+        _startTime = [NSDate date];
+        _retryCount = 0;
+        _requestCount = 0;
+        _pollingPaused = NO;
+        _pollingStopped = NO;
+        [self pollAfter:0 lastError:nil];
+        NSNotificationCenter *notificationCenter = [NSNotificationCenter defaultCenter];
+        [notificationCenter addObserver:self
+                               selector:@selector(restartPolling)
+                                   name:UIApplicationDidBecomeActiveNotification
+                                 object:nil];
+        [notificationCenter addObserver:self
+                               selector:@selector(restartPolling)
+                                   name:UIApplicationWillEnterForegroundNotification
+                                 object:nil];
+        [notificationCenter addObserver:self
+                               selector:@selector(pausePolling)
+                                   name:UIApplicationWillResignActiveNotification
+                                 object:nil];
+        [notificationCenter addObserver:self
+                               selector:@selector(pausePolling)
+                                   name:UIApplicationDidEnterBackgroundNotification
```

Comment:
I've modified it since the last review, and I think it has the correct behavior? When the app enters the background, we pause polling – if there's a request in-flight, it'll finish in the background, but no more requests will be made until the app becomes active.